File: prince-of-versions/src/main/java/co/infinum/princeofversions/RequirementsNotSatisfiedException.java
package co.infinum.princeofversions;

import java.util.HashMap;
import java.util.Map;

/**
 * Exception for cases where there is no update with requirements satisfied
 */
public class RequirementsNotSatisfiedException extends IllegalStateException {

    /**
     * Default metadata from root object in JSON file
     */
    private final Map<String, String> metadata;

    RequirementsNotSatisfiedException(final Map<String, String> metadata) {
        this.metadata = new HashMap<>(metadata);
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }

    @Override
    public String toString() {
        return "RequirementsNotSatisfiedException{"
            + "metadata=" + metadata
            + '}';
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/RequirementsNotSatisfiedException.kt
package co.infinum.princeofversions

import java.util.HashMap

/**
 * An exception thrown when there is no available update that satisfies the specified requirements.
 *
 * @param sourceMetadata The default metadata from the root object in the configuration file.
 * A defensive copy is made to ensure immutability.
 */
class RequirementsNotSatisfiedException(sourceMetadata: Map<String, String>) : IllegalStateException() {

    /**
     * A copy of the metadata from the update configuration.
     */
    val metadata: Map<String, String> = HashMap(sourceMetadata)

    override fun toString(): String {
        return "RequirementsNotSatisfiedException(metadata=$metadata)"
    }
}
